client/service/src/client/client.rs

```diff
@@ -896,39 +896,29 @@ where
 		operation.op.mark_finalized(BlockId::Hash(block), justification)?;
 
 		if notify {
-			let finalized =
-				route_from_finalized.enacted().iter().map(|elem| elem.hash).collect::<Vec<_>>();
+			let finalized = std::iter::once(last_finalized)
+				.chain(route_from_finalized.enacted().iter().map(|elem| elem.hash))
+				.collect::<Vec<_>>();
```

Comment:
```suggestion
			let finalized = 
				route_from_finalized.enacted().iter().map(|elem| elem.hash).collect::<Vec<_>>();
```

Reply:
I didn't included the `last_fnalized` (i.e. the previous last finalized block) it to prevent a double notification for that block:
- 1st notification when it is the last finalized block (the explicitly finalized one)
- 2nd notification when it s the fist implicitly finalized block

Indeed in the tests now a2 is repeated twice:
https://github.com/paritytech/substrate/blob/9ce46c344419d7dee25f01c24219d078ca459100/client/service/test/src/client/mod.rs#L1326-L1328

Genesis is the only block for which a notification was never sent since it is never explicitly finalized via a block import (implicitly finalized at startup)

In this way if blocks are (as it is usually the "normal" case) finalized one after the other then for each finalized block you'll always receive two finalizations (currently finalized one + prev finalized one)

Reply:
@davxy done :)